### DISCUSSION TITLE
Why is `System.SpanHelpers` || `new ReadOnlySpan<T>(ref T, int length)` internal?

### DISCUSSION THREAD

**Discussion by karakasa** [dotnet/runtime#92619] [General] [answered]
Recently I need to compare data stored in (`ref T` | `ref byte`) + length format, but I found there's no good way of doing that.

`new ReadOnlySpan<T>(ref T, int length)` is internal so I cannot create a `ROS`. The entire `SpanHelpers` class is internal so that I cannot use helper methods like `SequenceEqual<T>` (especially `SequenceEqual(ref byte, ref byte, nint)`), `SequenceCompareTo`, etc.

The methods don't seem unsafe at all. Is there a design consideration on this?

**Comment by stephentoub** [comment, marked as answer]
> new ReadOnlySpan<T>(ref T, int length) is internal so I cannot create a ROS

You want MemoryMarshal.CreateReadOnlySpan.

**Comment by huoyaoyuan** [comment]
Also note that, the `MemoryMarshal` or `CollectionsMarshal` class are considered "unsafe". Inappropriate usage can break the process. Other public APIs on `Span` or `MemoryExtensions` are considered safe.
